Scripts/Haptics/XRNodeHapticPulse.cs

```diff
@@ -0,0 +1,85 @@
+﻿namespace VRTK.Core.Haptics
+{
+    using UnityEngine;
+    using UnityEngine.XR;
+
+    /// <summary>
+    /// Creates a timed haptic pulse on a <see cref="XRNode"/> supported haptic device.
+    /// </summary>
+    public class XRNodeHapticPulse : HapticPulse
+    {
+        /// <summary>
+        /// The device to pulse.
+        /// </summary>
+        [Tooltip("The device to pulse.")]
+        public XRNode device = XRNode.LeftHand;
+        /// <summary>
+        /// The duration to pulse the device for.
+        /// </summary>
+        [Tooltip("The duration to pulse the device for.")]
+        public float duration = 0.005f;
+
+        /// <summary>
+        /// The haptic capabilities of the specified <see cref="XRNode"/>.
+        /// </summary>
+        protected HapticCapabilities caps = new HapticCapabilities();
+
+        /// <inheritdoc />
+        public override bool IsActive()
+        {
+            return base.IsActive();
+        }
+
+        /// <inheritdoc />
+        protected override void DoBegin()
+        {
+            Pulse(Intensity, duration, device);
+        }
+
+        /// <inheritdoc />
+        protected override void DoCancel()
+        {
+            Pulse(0, 0, device);
+        }
+
+        /// <summary>
+        /// Sends a pulse to the current device.
+        /// </summary>
+        /// <param name="intensity">The intensity to pulse with.</param>
+        /// <param name="duration">The duration to pulse for.</param>
+        /// <param name="node">The device to pulse on.</param>
+        protected virtual void Pulse(float intensity, float duration, XRNode node)
+        {
+            if (!InputDevices.GetDeviceAtXRNode(node).TryGetHapticCapabilities(out caps))
+            {
+                return;
+            }
+
+            if (caps.supportsImpulse)
+            {
+                InputDevices.GetDeviceAtXRNode(node).SendHapticImpulse(0, intensity, duration);
+            }
+            else if (caps.supportsBuffer)
+            {
+                byte[] clip = GeneratePulseBuffer(duration);
+                InputDevices.GetDeviceAtXRNode(node).SendHapticBuffer(0, clip);
+            }
+        }
+
+        /// <summary>
+        /// Generates a pulse buffer array.
+        /// </summary>
+        /// <param name="duration">The duration to pulse for.</param>
+        /// <returns>The buffer array containing the pulse data.</returns>
+        protected virtual byte[] GeneratePulseBuffer(float duration)
+        {
+            int clipCount = (int)(caps.bufferFrequencyHz * duration);
+            byte[] clip = new byte[clipCount];
+            for (int i = 0; i < clipCount; i++)
+            {
+                clip[i] = byte.MaxValue;
```

Comment:
This sets a buffer that is at full intensity at all times, see [these tests](https://github.com/Unity-Technologies/XRInputTests/tree/2018.3/Assets/Tests/HapticsTest/Scripts/ClipTests).

This is what we want in this case I think, so this is more of a comment than a review comment. Sadly we can't offer specific API to use this as the buffer support isn't guaranteed for everything and it will be hard to create a fallback that does the same by taking the user's buffer and then trying to call the impulse API to represent that buffer (if pulses are supported by the node that is).

Reply:
Intensity is a normalized value between 0 - 1. Can't we just multiple `byte.MaxValue` by `Intensity`? then if it's 1 you get the max value.?

`clip[i] = (byte)(byte.MaxValue * Intensity);`

Reply:
Oh perfect - I totally didn't come up with that because the `intensity` is not yet passed into this method as a parameter. Definitely do this (and remember to add the XML docs for this additional parameter - I often forget that one 😅).